tests/integration/application-charm/src/charm.py

```diff
@@ -0,0 +1,271 @@
+        if not hasattr(event, "relation"):
+            return
+
+        requires = self.relations.get(event.relation.name)
+        tls_ca = requires.fetch_relation_field(event.relation.id, "tls-ca")
+
+        if not tls_ca:
+            event.defer()  # We're waiting until we get a CA.
+            return
+
+        logger.info(f"writing cert to {CERT_PATH}.")
+        with open(CERT_PATH, "w") as f:
+            f.write(tls_ca)
+
+    def _run_request_action(self, event: ActionEvent, server):
+        """Generic request action handler to be used in the specific request action handlers."""
+        relation_id = event.params["relation-id"]
+        method = event.params["method"]
+        headers = event.params["headers"]
+        endpoint = event.params["endpoint"]
+        payload = event.params.get("payload", None)
+        if payload:
+            payload = payload.replace("\\", "")
+
+        protocol = "https"
+        if server.split("://") != [server]:
+            protocol, server = server.split("://")
+
+        requires = self._get_requires(event.params["relation-name"])
+        username = requires.fetch_relation_field(relation_id, "username")
+        password = requires.fetch_relation_field(relation_id, "password")
+
+        if not username or not password:
+            event.fail("Credentials are not accessible.")
+            return
+
+        host_addr, port = server.split(":")
+        logger.info(
+            f"Sending {method} request to {server}/{endpoint} with headers {headers} and payload {payload}"
+        )
+        try:
+            response = self.request(
+                method,
+                headers,
+                endpoint,
+                int(port),
+                username,
+                password,
+                host_addr,
+                payload,
+                protocol,
+            )
+        except OpenSearchHttpError as e:
+            response = [str(e)]
+        logger.info(response)
+
+        event.set_results({"results": json.dumps(response)})
+
+    # ==============
+    #  Action hooks
+    # ==============
+
+    def _on_run_db_request_action(self, event: ActionEvent):
+        """Send a database request."""
+        relation_id = event.params["relation-id"]
+        requires = self._get_requires(event.params["relation-name"])
+        servers = requires.fetch_relation_field(relation_id, "endpoints")
+        if not servers:
+            event.fail("No Opensearch servers to send the request to.")
+            return
+        server = servers.split(",")[0]
+
+        self._run_request_action(event, server)
+
+    def _on_run_dashboards_request_action(self, event: ActionEvent):
+        """Send a request to Opensearch Dashboards."""
+        server = event.params["server-uri"]
+        if not server:
+            event.fail("No Opensearch Dashboards servers to send the request to.")
+            return
+
+        self._run_request_action(event, server)
```

Comment:
...so the idea is that: `application` charm is simulating The User.

 - on one hand, the entity who is uploading data into OpenSearch
 - on the other hand, the physical user (may be the same as 'entity above') clicking around on Opensearch Dashboards.
 
 Therefore, unlike the "parent" `application` from Opensearch, here we distinguish whether an action goes against the DB or against the web UI.